src/storage-client/src/controller.rs

```diff
@@ -564,6 +566,26 @@ pub trait StorageController: Debug {
     /// Identifiers not present in `policies` retain their existing read policies.
     fn set_read_policy(&mut self, policies: Vec<(GlobalId, ReadPolicy<Self::Timestamp>)>);
 
+    /// Acquires and returns the desired read holds, advancing them to the since
+    /// frontier when necessary.
```

Comment:
I read this comment to mean that if I request a read hold before the `since`, I will instead receive one at the `since`? I think it would be more robust to instead return an error. This avoids bugs when the caller simply assumes that it got the requested read hold without checking.

Taking a step back, there are two scenarios I can think of when this API is called:
1. The caller wants a read hold at the earliest possible time, or just any read hold.
2. The caller wants a read hold at one specific time.

With this API, a caller can achieve (1) by `acquire_read_hold(id, Timestamp::MIN);` and (2) by `let hold = acquire_read_hold(id, requested_time); hold.try_downgrade(requested_time)?;`. But forgetting the check is easy.

How about we instead make the desired parameter an `Option` that when set to `Some` requests a precise read hold? Then the above usages would look like (1) `acquire_read_hold(id, None)` and (2) `acquire_read_hold(id, Some(requested_time))?`. Alternatively, two different methods, one that takes a desired frontier and one that doesn't would work as well. The latter could even be a wrapper around the former that just calls `try_downgrade`.

Reply:
Yeah, I've thought about this in the context of the adapter version of `acquire_read_holds` before. After a couple of recent refactorings, there's actually zero places in the code where we call that with a timestamp other than `MIN` _or_ with `precise = true`. So we can clean at least that up, I believe. cc @ggevay who initially added the `precise` parameter.

I'll push a commit that makes this here API more obvious.

Reply:
> there's actually zero places in the code where we call that with a timestamp other than `MIN` or with `precise = true`

I have to say I'm happy about this! Feel free to remove the `precise` stuff then!

Reply:
I agree, this is nicer! Still I'd prefer if we could remove the pitfall of `StorageController::acquire_read_holds` silently returning a hold that is later than requested. Also because now the `acquire_read_holds` and `acquire_read_hold_at_time` behave inconsistently (only one of them erroring on a too-early desired time).

AFAICT the only place this method is called externally is in adapter's `acquire_read_hold`, where the requested read hold is at `Timestamp::MIN`. So how about we change this method to one that always returns the `since` read holds?

```rust
    fn acquire_read_holds(
        &mut self,
        desired_holds: Vec<GlobalId>,
    ) -> Result<Vec<ReadHold<Self::Timestamp>>, ReadHoldError>;
```

Or, alternatively, if you think we might need a batched variant of `acquire_read_hold_at_time` in the future, this method could return a `ReadHoldError::SinceViolation` like `acquire_read_hold_at_time` does.